azure-iot-device/azure/iot/device/iothub/pipeline/pipeline_stages_iothub_mqtt.py

```diff
@@ -51,14 +51,19 @@ def _execute_op(self, op):
             else:
                 client_id = op.device_id
 
+            # For MQTT, the entire user agent string should be appended to the username field in the connect packet
+            # For example, the username may look like this without custom parameters:
+            # yosephsandboxhub.azure-devices.net/alpha/?api-version=2018-06-30&DeviceClientType=py-azure-iot-device%2F2.0.0-preview.12
+            # The customer user agent string would simply be appended to the end of this username, in URL Encoded format.
             query_param_seq = [
                 ("api-version", pkg_constant.IOTHUB_API_VERSION),
                 ("DeviceClientType", pkg_constant.USER_AGENT),
             ]
-            username = "{hostname}/{client_id}/?{query_params}".format(
+            username = "{hostname}/{client_id}/?{query_params}{optional_product_info}".format(
```

Comment:
>username = "{hostname}/{client_id}/?{query_params}{optional_product_info}".format( [](start = 12, length = 82)

What, no tests?  You need at least 1 test for a string that doesn't need URI encoding and a different test for a string that does need URI encoding.  If we don't have a test that validates the username without optional_product_info, we should have that, but I assume that is already there.  And, just for giggles, since this is user input, I might add tests for a few different types which aren't strings.  (integers at least, but I'm not sure about dictionary or list).  

And, now that I look at it, I don't see any URI-encoding in here.  I assume it's our responsibility to encode it?  Or does the user pass a string which is already encoded?  Either way, the responsibly party should be documented in create_from_connection_string, etc.

Reply:
This is a bug. Fixed this and will push soon. 

---
In reply to: [338265615](https://github.com/Azure/azure-iot-sdk-python/pull/368#discussion_r338265615) [](ancestors = 338265615)

Reply:
As for tests: yes no tests yet for the custom param yet. It's passing all our current tests without the product info stuff, but I'll be adding tests for validation soon. 

---
In reply to: [338272541](https://github.com/Azure/azure-iot-sdk-python/pull/368#discussion_r338272541) [](ancestors = 338272541,338265615)

Reply:
So I've added tests for test_pipeline_stages_iothub_mqtt now. Those tests manually set the pipeline_root configuration product_info to multiple values and validate that they are appended correctly after being url encoded. Those values include: empty string, normal string, array, dict, and number. All non-string values are cast to their string representation before being URL Encoded because that is how the code I've written handles non-string user input values.

---
In reply to: [338272889](https://github.com/Azure/azure-iot-sdk-python/pull/368#discussion_r338272889) [](ancestors = 338272889,338272541,338265615)